proxy/src/control/discovery.rs

```diff
@@ -199,27 +206,76 @@ where
-        let update = match up {
-            Ok(Async::Ready(Some(update))) => update,
-            Ok(Async::Ready(None)) => unreachable!(),
-            Ok(Async::NotReady) => return Ok(Async::NotReady),
-            Err(_) => return Err(()),
-        };
-
-        match update {
-            | Update::Insert(addr, meta) => {
-                let service = self.bind.bind(&addr)
-                    .map(|svc| meta.label(svc))
-                    .map_err(|_| ())?;
+        loop {
+            let up = self.rx.poll();
+            trace!("watch: {:?}", up);
+            let update = match up {
+                Ok(Async::Ready(Some(update))) => update,
+                Ok(Async::Ready(None)) => unreachable!(),
+                Ok(Async::NotReady) => return Ok(Async::NotReady),
+                Err(_) => return Err(()),
+            };
 
-                Ok(Async::Ready(Change::Insert(addr, service)))
-            },
-            // TODO: handle metadata changes by changing the labeling
-            // middleware to hold a `futures-watch::Watch` on the label value,
-            // so it can be updated.
-            Update::Remove(addr) => Ok(Async::Ready(Change::Remove(addr))),
+            match update {
+                Update::Insert(addr, meta) => {
+                    // Construct a watch for the `Labeled` middleware that will
+                    // wrap the bound service, and insert the store into our map
+                    // so it can be updated later.
+                    let (watch_labels, update_labels) =
+                        futures_watch::Watch::new(meta.metric_labels);
+                    self.metric_labels.insert(addr, update_labels);
+
+                    let service = self.bind.bind(&addr)
+                        .map(|svc| Labeled::new(svc, watch_labels))
+                        .map_err(|_| ())?;
+
+                    return Ok(Async::Ready(Change::Insert(addr, service)))
+                },
+                Update::ChangeMetadata(addr, meta) => match self
+                    .metric_labels
+                    .entry(addr)
+                {
+                    // Handle changes in metadata (currently, destination
+                    // labels by updating our `Store` for that service's set of
+                    // labels.
+                    Entry::Occupied(mut entry) => {
+                        let canceled = entry.get_mut()
+                            .poll_cancel()
+                            .map_err(|e| {
+                                error!("Watch: label poll_cancel: {:?}", e);
+                            });
+                        if canceled?.is_ready() {
+                            // If poll_cancel() returns `Ready`, then the
+                            // service bound to that address has been
+                            // dropped and we can safely remove it.
+                            let _ = entry.remove_entry();
+                            continue;
+                        };
```

Comment:
Why do we need to call poll_cancel here? It seems fine to me to just blindly store the label.

If anything, we should probably consider this case a more serious error (because we've received a ChangeMetadata after a Remove; but I don't think it's necessary to do this validation here.